spag/cli.py

```diff
@@ -34,66 +34,104 @@ def show_response(resp, show_headers):
         click.echo("ERROR: %s %s" % (str(resp.status_code), resp.reason))
         click.echo(resp.text)
 
+
+class Request(object):
+    """A Request object that looks more like request lib's Request but isn't"""
+
+    def __init__(self, req):
+        """
+        :param req: an instance of urllib2.Request
+        """
+        self.body = req.get_data()
+        self.headers = req.headers
+        self.method = req.get_method()
+        self.path_url = req.get_selector()
+        self.url = req.get_full_url()
+
+class Response(object):
+    """A Response that looks more requests lib's Response but isn't"""
+
+    def __init__(self, resp, req):
+        """
+        :param req: an instance of urllib2.Request or urllib2.HTTPError
+        :param resp: the result of urllib2.urlopen(...)
+        """
+        self.request = Request(req)
+        # we can only call resp.read() once
+        self.text = resp.read()
+        self.status_code = resp.code
+        self.headers = resp.headers
+        self.url = resp.url
+        # e.g. 'OK'
+        self.reason = resp.msg
+        self.ok = not isinstance(resp, urllib2.HTTPError)
+
+
+def do_request(method, resource, endpoint=None, header=None, data=None,
+               show_headers=False, remember_as=None):
+    req = urllib2.Request(
+        url = template.untemplate(endpoint + resource, shortcuts=True),
+        headers=header or {})
+    req.add_data(data)
+    req.get_method = lambda: method
+
+    try:
+        # urllib2's resp is file-like; we can only call resp.read() once
+        resp = urllib2.urlopen(req)
+        resp = Response(resp, req)
+    except urllib2.HTTPError as e:
+        resp = Response(e, req)
+    show_response(resp, show_headers)
+    remember_as = remember_as or method.lower()
+    remembers.SpagRemembers.remember_request(remember_as, resp)
+    remembers.SpagHistory.append(resp)
+
+    # how to do it with requests
```

Comment:
Using urllib2 was just an experiment. There are too many Python 2/3 differences to make it worthwhile. (`"TypeError: POST data should be bytes or an iterable of bytes. It cannot be of type str."`). Requests lib uses urllib3 which would probably do better, but importing urllib3 takes two-thirds the time of importing requests which doesn't really make a change.

Things we'd have to look into if did we move away from requests:

- http auth (basic/digest)
- ssl support/verification
- ???